### PR TITLE
[RDY] API: Implement nvim_win_set_buf

### PR DESCRIPTION
**Pull request by HiPhish**
The implementation is extremely crude and certainly missing tons of cleanup, but it works on paper.

### CODE REVIEW

**Comment by HiPhish** [comment]
The current implementation does change the buffer, but the user needs to switch to another window first to see the changes. I'm certain that just changing the buffer the window points to cannot be all there is to it. The `do_buffer` function is nearly 300 lines long, a good chunk of it is just figuring out what it needs to do based on the arguments, but there is also a lot of cleanup.

I have looked into the implementation of `nvim_set_current_buf`, but its use of `do_buffer` assumes that we want to set the buffer in the current window, not just any arbitrary window, so I cannot use its implementation.

As a side-note, `nvim_set_current_buf` could then be implemented on top of `nvim_win_set_buf` in VimScript as
~~~vim
function! nvim_set_current_buf(buffer)
    call nvim_win_set_buf(nvim_get_current_win(), a:buffer)
endfunction
~~~

**Comment by bfredl** [comment]
> I have looked into the implementation of nvim_set_current_buf, but its use of do_buffer assumes that we want to set the buffer in the current window, not just any arbitrary window, so I cannot use its implementation.

For better and worse, a lot of API (and recent vimL functionality) is implemented by changing curwin and/or curbuf, call an existing function that assumes curwin/curbuf, and then restoring the old current values. So I wouldn't be this quick to conclude that `do_buffer` can't be used.

**Comment by justinmk** [comment]
> but the user needs to switch to another window first to see the changes.

Most likely that indirectly runs do_buffer, accidentally doing the necessary fixups.

> nvim_set_current_buf, but its use of do_buffer assumes that we want to set the buffer in the current window

Why not set the current window temporarily, then unwind it? That's the usual way. Disabling autocommands etc makes it fast.

**Comment by HiPhish** [comment]
> For better and worse, a lot of API (and recent vimL functionality) is implemented by changing curwin and/or curbuf, call an existing function that assumes curwin/curbuf, and then restoring the old current values.

> Why not set the current window temporarily, then unwind it? That's the usual way.

And for a moment there I though I could get away from Vim's ritualistic code by going to the actual source. It's rituals all the way down, isn't it?

**Comment by HiPhish** [comment]
> Well, the insides can be ugly, but we hide the ugly behind a nice API. Better to get it right before worrying about elegance.

You are right. My problem with ritualistic code is that it is easy to miss part of the ritual and have things work fine for the most part, but not always. Take for example the updated code I just pushed, it works well, but the error handling is messed up. Looking at the function, it would have been better if I had just written it straight up in VimScript instead.
~~~vim
function! nvim_win_set_buf(win, buf)
    let old_win = nvim_get_current_win();
    call nvim_set_current_win(a:win)
    call nvim_set_current_buf(a:buf)
    call nvim_set_current_win(old_win)
endfunction
~~~
I had hoped that writing this directly in the source would avoid triggering fifty side effects from switching windows, but that would require digging even deeper into Neovim's core code and going down to the bare bytes. At this point I'm not sure if there is even a point in continuing with this anymore.

**Comment by bfredl** [comment]
> I had hoped that writing this directly in the source would avoid triggering fifty side effects from switching windows, but that would require digging even deeper into Neovim's core code and going down to the bare bytes. At this point I'm not sure if there is even a point in continuing with this anymore.

No need to be so defeatist this quickly. "Ritual" or not, other API functions have managed to avoid filthy side effects without too much raw byte banging. If we look at `nvim_win_set_option` (the arguably most side effect-full operation so far), and dig down the implementation in `api/private/helpers.c`, we find that it uses `switch_win`/`restore_win`, which is purposely built for this save/restore situation. It will probably work better that `goto_tabpage_win` which does implement user visible switching, and indeed would be equivalent to the vimscript visible functions.

**Comment by HiPhish** [comment]
@bfredl You are right, I guess I'm just in a bad mood and threw in the towel too early. `switch_win`/`restore_win` seems to do the trick, looking for comments on my implementation.

**Comment by justinmk** [comment]
Much better! Just needs a couple tests and address the TODO. Might want to add another sentence or two in the docstring explaining the motivation or an example.

**Comment by HiPhish** [comment]
I noticed something weird:
~~~vim
" There are two windows: 1000 and 1001, I am in window 1001 at the moment
call nvim_win_set_option(1000, 'stl', 'Hello world')
" The above changes the status line in the other window, as expected
call nvim_win_set_buf(1000, 2)
" The above changes the buffer, as expected
~~~
Now both windows show the same status line. Switching back to the original buffer restores my `Hello world` status line. It is as if the option is local to the buffer, not the window. Is this correct? The same happens if I stay inside the window I want to change and use `nvim_set_current_buf()`.

> Just needs a couple tests and address the TODO.

Sure thing, I just wanted to get this to work before applying the polish. What should be the API level? 6?

**Comment by justinmk** [comment]
>  It is as if the option is local to the buffer, not the window. Is this correct?

Not sure. How does it behave with VimL?

> What should be the API level? 6?

See NVIM_API_LEVEL in CMakeLists.txt

**Comment by HiPhish** [comment]
> Not sure. How does it behave with VimL?

That was VimL. Nevermind anyway, I looked into how options are set:
~~~c
// File src/nvim/private/helpers.c
// Omitting irrelevant code
static void set_option_value_for(char *key,
                                 int numval,
                                 char *stringval,
                                 int opt_flags,
                                 int opt_type,
                                 void *from,
                                 Error *err)
{
  switch (opt_type)
  {
    case SREQ_WIN:
      switch_win(&save_curwin, &save_curtab, (win_T *)from, win_find_tabpage((win_T *)from), false) 
      set_option_value_err(key, numval, stringval, opt_flags, err);
      restore_win(save_curwin, save_curtab, true);
      break;
    case SREQ_BUF:
      switch_buffer(&save_curbuf, (buf_T *)from);
      set_option_value_err(key, numval, stringval, opt_flags, err);
      restore_buffer(&save_curbuf);
      break;
  }
}
~~~
There is not really such a thing as a "window-local" option, `nvim_win_set_option(n, 'foo', 'bar')` is really just `nvim_buf_set_option(nvim_win_get_buf(n), 'foo', 'bar')`.

**Comment by bfredl** [comment]
Note the docs claim `statusline` is the one and only "global-local window option" (i e it doesn't be have like any other window option), so do expect weird behaviour with it, especially in the API. It seems `nvim_win_set_option(1000, 'stl', 'Hello world')` also sets the buffer local value, so when you visit the buffer in a new window that hasn't seen the buffer before, the local value gets copied. It might be worth trying with some other window option, like `winhl`.

**Comment by HiPhish** [comment]
> It might be worth trying with some other window option, like `winhl`.

I tried it and I see the same behaviour: the "window-local" options switches along with the buffer. It makes sense if you look at the code in my previous post: in both cases `set_option_value_err` is called with the same arguments, the only difference is the part of the ritual beforehand and after. For a window we switch to the window in question, thus also switching to its buffer, and for a buffer we switch straight to the buffer in question.

(I hate how Vim as taught me to press CTRL-W to delete the current word, I always end up closing a browser window after having almost finished a post)

**Comment by bfredl** [comment]
Ahyes, it likely behaves like `:setl` in both cases, but for a window option it should rather work like plain `:set`, except maybe for `stl` (which will remain an unhappy child in any case)

**Comment by HiPhish** [comment]
I have now written a test for the new function and cleaned up linter complaints. Requesting consideration for merging.

**Comment by bfredl** [comment]
History looks a bit weird. perhaps rebase the latest commit on master?

File: CMakeLists.txt [hunk 1]
```diff
@@ -113,7 +113,7 @@ set(NVIM_VERSION_PATCH 2)
 set(NVIM_VERSION_PRERELEASE "-dev") # for package maintainers
 
 # API level
-set(NVIM_API_LEVEL 5)         # Bump this after any API change.
+set(NVIM_API_LEVEL 6)         # Bump this after any API change.
```
Comment (bfredl) [inline CMakeLists.txt]: No need to bump this. One can always just use the latest level in a PR; if it needs to be bumped, CI  (or `make functionaltest` locally) will tell you.

**Comment by HiPhish** [comment]
Is the history better now? GitHub shows all those other commits from `master` for some reason.

**Comment by bfredl** [comment]
I think you merged instead, which doesn't remove these commits. Reset to
https://github.com/neovim/neovim/pull/9100/commits/c55287ad0e6d5c93f80fb09ad0b30867ac895166 again and use `git rebase -i master`. Hopefully it would only show one commit (implement API method), but if the rest remain, just delete those lines.

**Comment by bfredl** [comment]
Ref discussion https://github.com/neovim/neovim/pull/9100#issuecomment-429125151 and below.

**Comment by HiPhish** [comment]
I followed the instructions, but the history is still messed up. Could it be that I have messed up my `master` when I did `git pull upstream master`? I could do
~~~sh
git reset --hard 384770556ba02433904baf78f499b9bb48010040
~~~
on the `master` branch to get back to the last time before my `git pull` and then do things the right way (whatever the right way might be).

**Comment by justinmk** [comment]
@HiPhish I force-pushed to your branch. I used `git merge --squash` to get rid of the unnecessary commits:

    git checkout upstream/master
    git merge --squash refs/pull/upstream/9100
    git push -f https://github.com/HiPhish/neovim HEAD:api-win-set-buf

**Comment by HiPhish** [comment]
@justinmk Thank you. What is the Right Way™ of keeping up to data with `master`? This sort of thing keeps happening to me. Here is what I usually do:
~~~sh
git checkout master
git pull upstream master  # upstream is the official repository
git checkout my-feature-branch
git rebase master
~~~

File: test/functional/api/window_spec.lua [hunk 2]
```diff
@@ -45,6 +45,16 @@ describe('api/win', function()
...
 
+  describe('set_buf', function()
+    it('works', function()
+      nvim('command', 'new')
+      local windows = nvim('list_wins')
+      neq(window('get_buf', windows[2]), window('get_buf', windows[1]))
+      window('set_buf', windows[2], window('get_buf', windows[1]))
+      eq(window('get_buf', windows[2]), window('get_buf', windows[1]))
+	end)
+  end)
```
Comment (justinmk) [inline test/functional/api/window_spec.lua]: indentation

File: src/nvim/api/window.c [hunk 3]
```diff
@@ -33,6 +34,44 @@ Buffer nvim_win_get_buf(Window window, Error *err)
...
+
+  if (!win || !buf) {
+      return;
+  }
+
+  if (switch_win(&save_curwin, &save_curtab, win, tab, false) == FAIL) {
+    api_set_error(err,
+                  kErrorTypeException,
+                  "Failed to switch to window %d",
+                  window);
```
Comment (justinmk) [inline src/nvim/api/window.c]: should also test all of the failure modes (including the `!win || !buf` case above). In the past we've had hidden bugs until we added test cases for failure modes.

It's easy to test the error cases using `expect_err()`. See https://github.com/neovim/neovim/blob/8fd092f3ff15bf70f84ec0d716c5aaa2c7379fa1/test/functional/api/vim_spec.lua#L28                               
 for example.

Comment (HiPhish) [reply src/nvim/api/window.c]: I have added tests for invalid window- and buffer ID, but as you said, I have no idea how to make the swap intentionally fail.

**Comment by HiPhish** [comment]
*bump* Anything else that needs to be done?

**Comment by justinmk** [comment]
Merged after fixing indentation and API level. Thanks @HiPhish ,!

**Pull request closed without merging**